File: typescript/src/expiry.ts
/**
 * Carapace v0.2 — Card Expiry / TTL (TypeScript)
 *
 * Usage:
 *   import { makeExpiresAt, checkExpiry, isExpired, ExpiryStatus } from './expiry';
 *
 *   const expires = makeExpiresAt({ ttlHours: 24 });
 *   const status = checkExpiry(card);
 */

import type { CardLike } from './enforce';

// ── Types ────────────────────────────────────────────────────────────────────

export enum ExpiryStatus {
  VALID = 'valid',
  EXPIRED = 'expired',
  EXPIRING_SOON = 'expiring_soon',
  NO_EXPIRY = 'no_expiry',
}

export interface ExpiryCheckResult {
  passed: boolean;
  reason: string | null;
  expiresAt: string | null;
  status: ExpiryStatus;
  hoursRemaining?: number | null;
}

// ── Helpers ──────────────────────────────────────────────────────────────────

/**
 * Create an ISO 8601 expires_at string for card registration.
 *
 * makeExpiresAt({ ttlHours: 24 })
 * makeExpiresAt({ ttlDays: 90 })
 * makeExpiresAt({ absolute: '2026-12-31T23:59:59Z' })
 */
export function makeExpiresAt(opts?: {
  ttlHours?: number;
  ttlDays?: number;
  absolute?: string | Date;
}): string | null {
  if (!opts) return null;

  if (opts.absolute !== undefined) {
    if (opts.absolute instanceof Date) {
      return opts.absolute.toISOString();
    }
    return opts.absolute;
  }

  const hours = opts.ttlHours ?? 0;
  const days = opts.ttlDays ?? 0;
  const totalMs = (hours * 3600 + days * 86400) * 1000;

  if (totalMs <= 0) {
    if (hours === 0 && days === 0) return null;
    throw new Error('TTL must be positive');
  }

  return new Date(Date.now() + totalMs).toISOString();
}

/** Parse an expires_at string to a Date, or null. */
export function parseExpiresAt(value: string | Date | null | undefined): Date | null {
  if (value == null) return null;
  if (value instanceof Date) return value;
  const d = new Date(value);
  return isNaN(d.getTime()) ? null : d;
}

/**
 * Check a card's expiry status.
 *
 * @param warningThresholdMs - How far in advance to flag EXPIRING_SOON (default: 24h)
 * @param now - Override current time (for testing)
 */
export function checkExpiry(
  card: CardLike | { expires_at?: string | null },
  options?: {
    warningThresholdMs?: number;
    now?: Date;
  },
): ExpiryStatus {
  const raw = (card as any).expires_at ?? null;
  const expDate = parseExpiresAt(raw);

  if (!expDate) return ExpiryStatus.NO_EXPIRY;

  const now = options?.now ?? new Date();
  const threshold = options?.warningThresholdMs ?? 24 * 60 * 60 * 1000; // 24h

  if (now.getTime() > expDate.getTime()) return ExpiryStatus.EXPIRED;
  if (now.getTime() > expDate.getTime() - threshold) return ExpiryStatus.EXPIRING_SOON;
  return ExpiryStatus.VALID;
}

/** Quick boolean: is the card past its TTL? */
export function isExpired(
  card: CardLike | { expires_at?: string | null },
  now?: Date,
): boolean {
  return checkExpiry(card, { now }) === ExpiryStatus.EXPIRED;
}

/**
 * Time remaining in milliseconds. Null if no expiry.
 * Negative if already expired.
 */
export function timeRemaining(
  card: CardLike | { expires_at?: string | null },
  now?: Date,
): number | null {
  const raw = (card as any).expires_at ?? null;
  const expDate = parseExpiresAt(raw);
  if (!expDate) return null;
  return expDate.getTime() - (now ?? new Date()).getTime();
}

/**
 * Returns a result object for integration into verify().
 */
export function validateExpiryForVerify(
  card: CardLike | { expires_at?: string | null },
): ExpiryCheckResult {
  const status = checkExpiry(card);
  const expiresAt = (card as any).expires_at ?? null;

  if (status === ExpiryStatus.EXPIRED) {
    return { passed: false, reason: 'card_expired', expiresAt, status };
  }

  if (status === ExpiryStatus.EXPIRING_SOON) {
    const remaining = timeRemaining(card);
    return {
      passed: true,
      reason: null,
      expiresAt,
      status,
      hoursRemaining: remaining != null ? remaining / (1000 * 3600) : null,
    };
  }

  return { passed: true, reason: null, expiresAt, status };
}

File: v030/expiry.py
"""
Carapace v0.2 — Card Expiry / TTL

Adds expires_at support to the agent card lifecycle.
One schema field, huge enterprise signal (SOC 2, FedRAMP).

This module provides:
- TTL validation for verify() and verifyLocal() integration
- Helper functions for creating cards with expiry
- Expiry status checking without full verification
"""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
from enum import Enum
from typing import Any


class ExpiryStatus(Enum):
    """Card expiry states."""
    VALID = "valid"              # Not expired, or no expiry set
    EXPIRED = "expired"          # Past expires_at
    EXPIRING_SOON = "expiring_soon"  # Within warning threshold
    NO_EXPIRY = "no_expiry"      # No expires_at field (v0.1.1 behavior)


# ── Expiry Helpers ────────────────────────────────────────────────────────────

def make_expires_at(
    ttl_hours: int | None = None,
    ttl_days: int | None = None,
    absolute: datetime | str | None = None,
) -> str | None:
    """
    Produce an ISO 8601 expires_at value for card registration.

    Usage:
        # Relative TTL
        expires = make_expires_at(ttl_hours=24)
        expires = make_expires_at(ttl_days=90)

        # Absolute deadline
        expires = make_expires_at(absolute="2026-12-31T23:59:59Z")
        expires = make_expires_at(absolute=datetime(2026, 12, 31, tzinfo=timezone.utc))

    Returns ISO 8601 string or None.
    """
    if absolute is not None:
        if isinstance(absolute, str):
            return absolute
        if isinstance(absolute, datetime):
            if absolute.tzinfo is None:
                absolute = absolute.replace(tzinfo=timezone.utc)
            return absolute.isoformat().replace("+00:00", "Z")
        raise TypeError(f"absolute must be str or datetime, got {type(absolute)}")

    if ttl_hours is not None or ttl_days is not None:
        delta = timedelta(
            hours=ttl_hours or 0,
            days=ttl_days or 0,
        )
        if delta.total_seconds() <= 0:
            raise ValueError("TTL must be positive")
        expires = datetime.now(timezone.utc) + delta
        return expires.isoformat().replace("+00:00", "Z")

    return None  # No expiry


def parse_expires_at(value: Any) -> datetime | None:
    """Parse an expires_at value to a timezone-aware datetime, or None."""
    if value is None:
        return None
    if isinstance(value, datetime):
        if value.tzinfo is None:
            return value.replace(tzinfo=timezone.utc)
        return value
    if isinstance(value, str):
        try:
            dt = datetime.fromisoformat(value.replace("Z", "+00:00"))
            if dt.tzinfo is None:
                dt = dt.replace(tzinfo=timezone.utc)
            return dt
        except ValueError:
            return None
    return None


def check_expiry(
    card: Any,
    *,
    warning_threshold: timedelta = timedelta(hours=24),
    now: datetime | None = None,
) -> ExpiryStatus:
    """
    Check a card's expiry status without performing full verification.

    Args:
        card: AgentCard or dict with optional expires_at field.
        warning_threshold: How far in advance to flag EXPIRING_SOON.
        now: Override current time (for testing).

    Returns:
        ExpiryStatus enum value.
    """
    raw = (
        card.get("expires_at") if isinstance(card, dict)
        else getattr(card, "expires_at", None)
    )
    expires_dt = parse_expires_at(raw)

    if expires_dt is None:
        return ExpiryStatus.NO_EXPIRY

    current = now or datetime.now(timezone.utc)

    if current > expires_dt:
        return ExpiryStatus.EXPIRED

    if current > (expires_dt - warning_threshold):
        return ExpiryStatus.EXPIRING_SOON

    return ExpiryStatus.VALID


def is_expired(card: Any, *, now: datetime | None = None) -> bool:
    """Quick boolean check: is the card past its TTL?"""
    return check_expiry(card, now=now) == ExpiryStatus.EXPIRED


def time_remaining(card: Any, *, now: datetime | None = None) -> timedelta | None:
    """
    Time remaining before card expires. None if no expiry set.
    Negative timedelta if already expired.
    """
    raw = (
        card.get("expires_at") if isinstance(card, dict)
        else getattr(card, "expires_at", None)
    )
    expires_dt = parse_expires_at(raw)

    if expires_dt is None:
        return None

    current = now or datetime.now(timezone.utc)
    return expires_dt - current


# ── Verify Integration ────────────────────────────────────────────────────────
# These functions are meant to be called from within verify() / verifyLocal()
# as additional checks layered on top of signature verification.

def validate_expiry_for_verify(card: Any) -> dict[str, Any]:
    """
    Returns a dict to merge into VerifyResult when expiry applies.

    Usage inside verify():
        result = verify_signature(card)
        expiry_result = validate_expiry_for_verify(card)
        if not expiry_result["passed"]:
            return VerifyResult(verified=False, reason=expiry_result["reason"])
    """
    status = check_expiry(card)

    if status == ExpiryStatus.EXPIRED:
        return {
            "passed": False,
            "reason": "card_expired",
            "expires_at": getattr(card, "expires_at", None),
            "status": "expired",
        }

    if status == ExpiryStatus.EXPIRING_SOON:
        remaining = time_remaining(card)
        return {
            "passed": True,
            "reason": None,
            "expires_at": getattr(card, "expires_at", None),
            "status": "expiring_soon",
            "hours_remaining": remaining.total_seconds() / 3600 if remaining else None,
        }

    return {
        "passed": True,
        "reason": None,
        "expires_at": getattr(card, "expires_at", None),
        "status": status.value,
    }
